ArdusatSDK.h

```diff
@@ -49,152 +66,407 @@ typedef enum {
  */
 struct _data_header_v1 {
 	unsigned char unit;       // unit (standard) of the values (e.g. meter, m/s^2, etc.)
 	unsigned char sensor_id;  // id of the sensor that generated this data
 	unsigned long timestamp;  // millis for timestamping the data
 };
-
 typedef struct _data_header_v1 _data_header_t;
 
 /**
- * Data types are used to store/read sensor data according to the appropriate value type
+ * Get a string representation of a unit constant
  */
-typedef struct {
-	_data_header_t header;
-	float x,y,z;
-} acceleration_t;
-
-typedef struct {
-	_data_header_t header;
-	float x,y,z;
-} magnetic_t;
-
-typedef struct {
-	_data_header_t header;
-	float x,y,z;
-} gyro_t;
-
-typedef struct {
-	_data_header_t header;
-	float t;
-} temperature_t;
-
-typedef struct {
-	_data_header_t header;
-	float lux;
-} luminosity_t;
-
-typedef struct {
-	_data_header_t header;
-	float uvindex;
-} uvlight_t;
-
-typedef struct {
-	_data_header_t header;
-	float roll, pitch, heading;
-} orientation_t;
-
-typedef struct {
-	_data_header_t header;
-	float pressure;
-} pressure_t;
-
-#ifdef __cplusplus
-extern "C" {
-#endif
+const char * unit_to_str(unsigned char unit);
 
 /**
- * Where all of the sensor data is kept before being printed
+ * creates a string representation of the data in a CSV format that can be used with
+ * http://experiments.ardusat.com to visualize and log data.
  */
-extern int OUTPUT_BUF_SIZE;
-extern char * _output_buffer;
-char * _getOutBuf();
-void _resetOutBuf();
+const char * valuesToCSV(const char *sensorName, unsigned long timestamp, int numValues, ...);
+const char * valueToCSV(const char *sensorName, float value, unsigned long timestamp=0);
 
 /**
- * Get a string representation of a unit constant
+ * creates a string representation of the data in a JSON format that can be used with
+ * http://experiments.ardusat.com to visualize and log data.
+ *
+ * Format is:
+ * ~{"sensorName": "name", "unit": "C", "value": 35.3}|
  */
-const char * unit_to_str(unsigned char);
+const char * valuesToJSON(const char *sensorName, unsigned char unit, int numValues, ...);
+const char * valueToJSON(const char *sensorName, unsigned char unit, float value);
 
-/**
- * setup and read functions are used to initialize sensors and read data from them.
- */
-//TODO: expand setup functions with options
-boolean beginTemperatureSensor();
-void readTemperature(temperature_t & temp);
 
-boolean beginInfraredTemperatureSensor();
-void readInfraredTemperature(temperature_t & temp);
+/**************************************************************************//**
+ * @class Sensor
+ * @defgroup sensor
+ * @brief Base Sensor class that all Sensors inherit from
+ *****************************************************************************/
+class Sensor {
+  protected:
+    virtual boolean initialize(void) = 0;
+    void initializeHeader(sensor_id_t sensor_id, data_unit_t unit, const char name[] PROGMEM);
```

Comment:
But the `sensor_id_t`, i thought, was used to identify the physical board the value was coming from. Should there be more possible `enum` values? Because as of right now, I don't really understand how it's not a strict 1:1 mapping. I suppose you could allow the user to manually set that a temperature reading is coming from `SENSORID_BMP180`? But that doesn't really make sense.